src/ngx_http_lua_pipe.c

```diff
@@ -688,7 +688,7 @@ ngx_http_lua_ffi_pipe_spawn(ngx_http_lua_ffi_pipe_proc_t *proc,
         /* close listening socket fd */
         ls = ngx_cycle->listening.elts;
         for (i = 0; i < ngx_cycle->listening.nelts; i++) {
-            if (ngx_close_socket(ls[i].fd) == -1) {
+            if (ls[i].fd != -1 && ngx_close_socket(ls[i].fd) == -1) {
                 ngx_log_error(NGX_LOG_EMERG, ngx_cycle->log, ngx_socket_errno,
```

Comment:
As a side note, maybe these logs shouldn't have such a high severity. A warning maybe, but an error here should not impact the rest of the code path, so anything above that seems harder to justify, especially so `emerg`.